### PR TITLE
positive_definite?, positive_semidefinite?, full_rank?, rank_deficient?, get_rank 

### PR DESCRIPTION
**Pull request by johnmarinelli**
Hi!  Here is my first draft for https://github.com/SciRuby/nmatrix/issues/411.

One concern:
Performing the definite check gets slow when matrix sizes get into the thousands.  i'm currently working on implementing gershgorin's circle theorem to see if this resolves the issue.

Let me know what you think.  Thanks!

### CODE REVIEW

**Comment by mohawkjohn** [comment]
Looks like Travis failed the build.

File: lib/nmatrix/nmatrix.rb [hunk 1]
```diff
@@ -1124,5 +1179,3 @@ def maybe_get_noncomplex_default_value(to_dtype) #:nodoc:
 require_relative './math.rb'
 require_relative './enumerate.rb'
 
-require_relative './version.rb'
-require_relative './blas.rb'
```
Comment (wlevine) [inline lib/nmatrix/nmatrix.rb]: Did you mean to remove these? I think this is why the tests are failing.

Comment (johnmarinelli) [inline lib/nmatrix/nmatrix.rb]: @wlevine you're completely right, thanks!

**Comment by johnmarinelli** [comment]
there seems to be errors with any ruby version that isn't 2.2.  i'll be looking into it.  one thing I noticed is that I may need to explicitly cast a scalar value from float64 to float32, and complex128 to complex64 - is there a way to do that?  @mohawkjohn

**Comment by mohawkjohn** [comment]
No. Floats are always stored by Ruby as doubles, and likewise for complex.

As a side note, is there anyway we can overload #rank instead of calling it #get_rank?

**Comment by mohawkjohn** [comment]
Still appears to be failing.

**Comment by mohawkjohn** [comment]
Oh, you mean cast on the C side? Yes. There should be examples throughout.

**Comment by johnmarinelli** [comment]
alright - i'll look into it!  and i think overloading rank is a good idea.

**Pull request closed without merging**